src/groth16/aggregate/accumulator.rs

```diff
@@ -2,13 +2,126 @@
 use crate::bls::Engine;
 #[cfg(feature = "blst")]
 use blstrs::PairingCurveAffine;
-use core::default::Default;
+
+use crossbeam_channel::{bounded, Receiver, Sender};
 use ff::{Field, PrimeField};
 use groupy::{CurveAffine, CurveProjective};
 #[cfg(feature = "pairing")]
 use paired::{Engine, PairingCurveAffine};
-use rand::rngs::OsRng;
 use rayon::prelude::*;
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering::SeqCst},
+    Arc, Mutex,
+};
+
+#[derive(Debug)]
+pub struct PairingChecks<E: Engine, R: rand::RngCore + Send> {
+    /// Circuit breaker to allow canceling all checks and marking the whole check as failed.
+    valid: Arc<AtomicBool>,
+    merge_send: Sender<PairingCheck<E>>,
+    valid_recv: Receiver<bool>,
+    /// Random number generator used for generating the random coefficients.
+    rng: Mutex<R>,
+    /// Ensures that the non randomized check is only added exactly once.
+    non_random_check_done: AtomicBool,
+}
+
+impl<E: Engine, R: rand::RngCore + Send> PairingChecks<E, R> {
+    pub fn new(rng: R) -> Self {
+        let (merge_send, merge_recv): (Sender<PairingCheck<E>>, Receiver<PairingCheck<E>>) =
+            bounded(10);
+        let (valid_send, valid_recv) = bounded(1);
+
+        let valid = Arc::new(AtomicBool::new(true));
+        let valid_copy = valid.clone();
+
+        rayon::spawn(move || {
+            let mut acc = PairingCheck::new();
+            while let Ok(tuple) = merge_recv.recv() {
+                // only do work as long as we know we are still valid
+                if valid_copy.load(SeqCst) {
+                    acc.merge(&tuple);
+                }
+            }
+            if valid_copy.load(SeqCst) {
+                valid_send.send(acc.verify()).expect("failed to send");
+            }
+        });
+
+        PairingChecks {
+            valid,
+            merge_send,
+            valid_recv,
+            rng: Mutex::new(rng),
+            non_random_check_done: AtomicBool::new(false),
+        }
+    }
+
+    /// Fails the whole check.
+    pub fn invalidate(&self) {
+        self.valid.store(false, SeqCst);
+    }
+
+    fn merge_pair(&self, result: E::Fqk, exp: E::Fqk, must_randomize: bool) {
+        self.merge(PairingCheck::from_pair(result, exp), must_randomize);
+    }
+
+    fn merge_miller_one(&self, result: E::Fqk, must_randomize: bool) {
+        self.merge(PairingCheck::from_miller_one(result), must_randomize);
+    }
+
+    pub fn merge_pairing_equation(&self, left: Vec<E::Fqk>, pair: (E::Fqk, E::Fqk)) {
```

Comment:
Could we rename it to `merge_miller_outputs` in contrast to the other one that takes the inputs to the miller loop ? Also the last argument should be a single element, not a pair, it's the "right" side of equation of the form `e(A,B)e(C,D)... = T`

Reply:
Or maybe we should rename the function more explicitly geared towards their usage like : `randomize_and_merge`and `merge`

Reply:
feel free to rename and merge, so we have this on the main branch